File: 032-base64/base64.go
package main

import (
	"errors"
	"fmt"
	"io"
	"os"
)

/* Arrays cannot be constants in Go. */
var Base64Chars [64]uint8 = [64]uint8{'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '+', '/'}

func Base64Encode(input []uint8, buf []uint8, last bool, output io.Writer) (int, error) {
	chunks, rem := len(input)/3, len(input)%3
	buf = buf[0:0]
	for c := 0; c < chunks; c++ {
		b1, b2, b3 := input[c*3+0], input[c*3+1], input[c*3+2]
		buf = append(buf, Base64Chars[(b1>>2)&0x3f])
		buf = append(buf, Base64Chars[((b1<<4)&0x30)|((b2>>4)&0x0f)])
		buf = append(buf, Base64Chars[((b2<<2)&0x3c)|((b3>>6)&0x03)])
		buf = append(buf, Base64Chars[(b3&0x3f)])
	}

	if !last {
		_, err := output.Write(buf)
		return chunks * 3, err
	}

	if rem == 1 {
		b1 := input[chunks*3+0]
		buf = append(buf, Base64Chars[(b1>>2)&0x3f])
		buf = append(buf, Base64Chars[(b1<<4)&0x30])
		buf = append(buf, '=')
		buf = append(buf, '=')
	} else if rem == 2 {
		b1, b2 := input[chunks*3+0], input[chunks*3+1]
		buf = append(buf, Base64Chars[(b1>>2)&0x3f])
		buf = append(buf, Base64Chars[((b1<<4)&0x30)|((b2>>4)&0x0f)])
		buf = append(buf, Base64Chars[(b2<<2)&0x3c])
		buf = append(buf, '=')
	}

	_, err := output.Write(buf)
	return chunks*3 + rem, err
}

func main() {
	bufcap := 4096
	buf := make([]uint8, bufcap)
	tmpbuf := make([]uint8, bufcap+bufcap/2)
	startoffset := 0
	for {
		done := false
		n, err := os.Stdin.Read(buf[startoffset:])
		if err != nil {
			if !errors.Is(err, io.EOF) {
				fmt.Fprintf(os.Stderr, "error while reading from stdin: %s", err.Error())
				os.Exit(1)
				return
			}
			done = true
		}

		m, err := Base64Encode(buf[0:startoffset+n], tmpbuf, done, os.Stdout)
		if err != nil {
			fmt.Fprintf(os.Stderr, "error while writing to stdout: %s", err.Error())
			os.Exit(1)
			return
		}

		if done {
			os.Stdout.Close()
			break
		}

		startoffset = (startoffset + n) - m
		for i := 0; i < startoffset; i++ {
			buf[i] = buf[m+i]
		}
	}
}

File: 032-base64/base64.rs
#![feature(slice_as_chunks)]

use std::io::Read;
use std::io::Write;

const BUF_SIZE: usize = 4096;

const BASE64_CHARS: [u8; 64] = [
    b'A', b'B', b'C', b'D', b'E', b'F', b'G', b'H', b'I', b'J', b'K', b'L', b'M', b'N', b'O', b'P', b'Q', b'R', b'S', b'T', b'U', b'V', b'W', b'X', b'Y', b'Z', b'a', b'b', b'c', b'd', b'e', b'f', b'g', b'h', b'i', b'j', b'k', b'l', b'm', b'n', b'o', b'p', b'q', b'r', b's', b't', b'u', b'v', b'w', b'x', b'y', b'z', b'0', b'1', b'2', b'3', b'4', b'5', b'6', b'7', b'8', b'9', b'+', b'/'
];

fn base64_encode(buf: &[u8], out: &mut impl std::io::Write) -> std::io::Result<()> {
    assert!(buf.len() % 3 == 0);
    let mut outbuf = Vec::with_capacity(BUF_SIZE * 2);
    for [b1, b2, b3] in buf.as_chunks::<3>().0 {
        outbuf.push(BASE64_CHARS[((b1 >> 2) & 0x3f) as usize]);
        outbuf.push(BASE64_CHARS[(((b1 << 4) & 0x30) | ((b2 >> 4) & 0x0f)) as usize]);
        outbuf.push(BASE64_CHARS[(((b2 << 2) & 0x3c) | ((b3 >> 6) & 0x03)) as usize]);
        outbuf.push(BASE64_CHARS[(b3 & 0x3f) as usize]);
    }

    out.write_all(&outbuf)
}

fn base64_encode_finish(buf: &[u8], out: &mut impl std::io::Write) -> std::io::Result<()> {
    assert!(buf.len() < 3 && buf.len() != 0);
    let mut outbuf = Vec::with_capacity(BUF_SIZE * 2);

    if buf.len() == 1 {
        let b1: u8 = buf[0];
        outbuf.push(BASE64_CHARS[((b1 >> 2) & 0x3f) as usize]);
        outbuf.push(BASE64_CHARS[((b1 << 4) & 0x30) as usize]);
        outbuf.push(b'=');
        outbuf.push(b'=');
    } else if buf.len() == 2 {
        let b1: u8 = buf[0];
        let b2: u8 = buf[1];
        outbuf.push(BASE64_CHARS[((b1 >> 2) & 0x3f) as usize]);
        outbuf.push(BASE64_CHARS[(((b1 << 4) & 0x30) | ((b2 >> 4) & 0x0f)) as usize]);
        outbuf.push(BASE64_CHARS[((b2 << 2) & 0x3c) as usize]);
        outbuf.push(b'=');
    }

    out.write_all(&outbuf)
}

fn main() {
    let mut stdin = std::io::stdin().lock();
    let mut stdout = std::io::stdout().lock();
    let mut rem: usize = 0;
    let mut buf: [u8; BUF_SIZE] = [0u8; BUF_SIZE];
    loop {
        let n = match stdin.read(&mut buf[rem..]) {
            Ok(0) => break,
            Ok(n) => n + rem,
            Err(e) => {
                eprintln!("I/O error: {}", e);
                std::process::exit(1);
            }
        };

        let chunks = n / 3;
        rem = n % 3;
        base64_encode(&buf[0..(chunks * 3)], &mut stdout).expect("I/O error");
        for i in 0..rem {
            buf[i] = buf[(chunks * 3) + i];
        }
    }
    if rem != 0 {
        base64_encode_finish(&buf[0..rem], &mut stdout).expect("I/O error");
    }
    stdout.flush().expect("I/O error");
}
